### PR TITLE
Split cmmgen into generic cmm helpers and clambda-specific transformations

### PR DESCRIPTION
**Pull request by lthls**
This proposal splits the file `cmmgen.ml` into a minimal traversal of the `Clambda.ulambda` terms, which remains in the original file, and a new file called `cmm_helpers.ml` which contains most of the helper functions that were part of `Cmmgen` before, and some additional functions that were extracted from individual `match` clauses of the translation functions.

The idea for this proposal came when the flambda team started considering targeting `Cmm` directly as the output of its optimization rounds, instead of `Clambda` as the current version does.
This in turn prompted us to consider cleaning up `Cmmgen`, and the result of this cleanup is this PR.
The addition of an `.mli` file also contributes to documenting a part of the compiler that was until then lacking in documentation.
I've also asked @SimonJF for his opinion, as he is interested in targeting `Cmm` in his WebAssembly project, and he welcomes the change.

The pull request contains three parts:
- A rebased version of #1579 as it introduces some helpful simplifications.
- A single commit, labeled "Make Cmm generation independent of evaluation order", which is used to check the correctness of the following commits (but can be removed in the final version).
- The rest of the commits represent the refactoring itself. History isn't very important, but the code compiles at each point, and reviewing commit by commit can be easier.

In addition to the usual testsuite coverage, the correctness of the last part of the PR can be checked by comparing the generated object files byte-wise: I have checked on my machine, as a sample, that the file `ocamlc.opt` generated after commit 9addfa6 is equal, byte for byte, to the one generated by this PR, after commit 16bbfbe.
(The alternate way to view this is that this PR changes the order in which sub-expressions are explored, and this in turn changes the order in which global constants are aggregated. The extra determinism commit gives us a way to ensure that this is the only thing that gets modified.)

Some design choices were made to keep everything related to the unboxing strategy in `Cmmgen`, as other users of `Cmm_helpers` might have different, incompatible strategies. Optimization of conditionals also stayed there for the same reason.
I also tried to keep most of the helper functions stateless, but unfortunately some of the functions rely on an implicit environment in `Compilenv`, and I have decided not to abstract over it or replace it as I felt it would be more invasive.

Finally, there are a few `TODO`s remaining in the `cmm_helpers.mli` file, on some functions that do not need to be exported in the current state but either I felt other users might need it, or I wrote a bit of documentation for it and felt the mli was the right place to put it.

### CODE REVIEW

**Comment by mshinwell** [comment]
I have undertaken to review this fine piece of work once an initial discussion has taken place here.

File: asmcomp/clambda_primitives.mli [hunk 1]
```diff
@@ -0,0 +1,141 @@
+
```
Comment (alainfrisch) [inline asmcomp/clambda_primitives.mli]: Missing copyright headers (also for the .ml).

Comment (lthls) [reply asmcomp/clambda_primitives.mli]: This is part of #1579 by @chambart ; if this one looks more likely to be merged, we'll fix it here, but I wouldn't mind if #1579 was fixed and merged first on its own.

**Comment by mshinwell** [comment]
@lthls Please fix the check-typo failures, then this should go through CI.

**Comment by mshinwell** [comment]
@lthls And also a Changes entry for the Cmmgen change itself.

**Comment by lthls** [comment]
I'm not sure if it's a problem on my side, but running `tools/check-typo` outputs a bad header warning for every single source file it checks, which drowns out the real warnings. I can reproduce the issue on a clean checkout of trunk. In case it's relevant, `awk` is `mawk` on my system.

**Comment by lthls** [comment]
The travis failure is one of the jobs that was interrupted before it even started. The previous commit (which differs only by the changes entry) passes all travis tests.

**Comment by damiendoligez** [comment]
> I'm not sure if it's a problem on my side, but running tools/check-typo outputs a bad header warning for every single source file it checks, which drowns out the real warnings. I can reproduce the issue on a clean checkout of trunk. In case it's relevant, awk is mawk on my system.

You could try to launch `check-typo -missing-header` to ignore headers. Meanwhile I'll see if I can make it compatible with mawk.

**Comment by lthls** [comment]
Thanks for the tip, although the check-typo related issues have already been fixed.
If you happen to have any suggestions on how to move this PR forward, they would be very welcome.
@mshinwell and I would like to move forward (rebase, review & cleanup), but we're a bit wary of spending time on this without anybody else approving the general idea.

**Comment by damiendoligez** [comment]
> Meanwhile I'll see if I can make it compatible with mawk.

It would be pretty hard to port the script to mawk because the regexp language supported by mawk is much smaller than gawk's regexp language.

**Comment by lthls** [comment]
What about simply disabling the missing headers check when mawk is detected (with a warning) ?
The false positives were surprising, but requiring to install gawk to run the checks properly doesn't seem too constraining.

**Comment by mshinwell** [comment]
Gosh, `check-typo` is a popular topic of conversation these days.
Now, would anyone care to opine on the technical substance of this patch? :)

**Comment by mshinwell** [comment]
@xavierleroy Are you happy with the split in logic proposed here?  I'm ready to review this, but I would like your sign-off first.  (This one is taking some effort to keep rebased.)

**Comment by xavierleroy** [comment]
I'm OK with the general approach.  The diff is too large for me to review.

**Comment by mshinwell** [comment]
I will review this.  However we're going to get the GPRs that affect this code for the gdb work done first.

**Comment by mshinwell** [comment]
I have reviewed this.  I think with reasonable certainty it is correct.  It also seems a significant improvement on the current code.  Some minor points:
- There is a reference to the "asmrun" directory which should be "runtime".
- The "offset int" case should maybe use `add_int_caml`.
- There is a redundant comment about `pfield_computed` which can be deleted.

Since @xavierleroy has indicated is happy with the general approach, I am going to approve this for merging.  Please merge after the above has been fixed.

**Review by mshinwell — APPROVED** [review]

**Comment by mshinwell** [comment]
@lthls Apologies, I've realised I forgot to re-read the `.mli` file.  I will do this today, please hold on before merging.  The perils of not having a proper code review system...

**Comment by mshinwell** [comment]
Comments on the new `.mli` file:
- "Boxed integers headers" -> "Boxed integer headers"
- "functionally" spelt incorrectly, line 117
- "addr_array_length" and "float_array_length" should maybe be renamed, as the names are potentially misleading given the semantics
- "From the implementation of" (line 234) -> "For" (and reformat comment)
- Line 262, extraneous blank line
- Lines 306 to 309 would be better in cmm.mli.
- Line 327: "curryification" -> "currying wrapper" (likewise for "tuplification")
- Line 361: "into a block" is unnecessary
- Line 417 (etc.): "caml integer" -> "OCaml integer"
- Line 420 and 423: "fixed number" -> "constant integer"
- Lines 430 and 434 should agree on the wording.
- Line 465: "String" -> "Strings"
- Line 539: "From transl" isn't helpful
- Line 544: "of a symbol" -> "of a function via a symbol"
- Line 547: missing full stop
- Line 548: "load" -> "loaded"
- Line 549: "uses" -> "used"?  It isn't clear if this semantics still applies just from reading this comment.
- Line 596: move "respectively" after "positions"
- Line 619: "block" -> "blocks"

The only other comment is that I think labelled arguments could be used to good effect here.  This could mostly be done as a second patch in the future, although it isn't clear if it's worth disturbing the abstractions such as "ternary_primitive" for this.  However I think it may be worth adding labelled arguments in the cases of `float_array_ref` and `unboxed_float_array_ref` right now, which will mean that part of the comment can be deleted, and potentially save an error regarding (un)boxing.

OK to merge with those changes.

**Comment by lthls** [comment]
I've handled most of your comments. For `addr_array_length` and `float_array_length`, I've added a suffix `_shifted` that means that I had to change some amount of code to stay within 80 characters. I'm actually considering removing them from the `.mli`, as they're not used outside `Cmm_helpers` at the moment.
For the `bswap` comments, I haven't changed anything since the difference has a purpose: `bbswap` operates on integer types that are boxed in general, and the exposed function advertises that one must first unbox the values before calling it; while for `bswap16`, the type of the argument is `int`, so one has to first untag the argument.

**Comment by lthls** [comment]
With #8735 merged, there aren't many outstanding PRs which touch `Cmmgen`, so I think this is a good time to merge this one.
If nobody objects (and CI is green), I'll merge this tomorrow.

**Comment by lthls** [comment]
Merged. Thanks @mshinwell for the review, and apologies in advance to the people who had incoming patches on `Cmmgen` and will need some careful rebasing.

**Pull request merged**